### PR TITLE
Build process: use Celery handlers

### LINKED ISSUE #3984: Use Celery Task event methods instead of updating at context manager level
Raising this because of #3864 and some issues we hit today.

Currently we're using context managers to try to grab and report all the exceptions that we can. We could actually be using `Task.on_failure`[0], `Task.on_success`[1], and `Task.on_retry`[2] though. Instead of grabbing and reporting exceptions to users through the context manager, instead just bubble these up to Celery and update the build state at task failure/sucess.

This would be tried in steps:

* First, break out the update build code from the context manager. This will live in a utility function so it can be shared. Input is build_pk and the build state
* Next, make an exception subclass that can bubble up to Celery. That is, one that is not reported to users, but raises up through the context managers
* Make VCS locking a bubble up exception subclass, use `autoretry_on` to retry the task. (Do we mark as build failed here? Possible, but maybe we get away from that pattern)
* Get rid of generic handling of errors outside buildenvs and use `on_failure` for this

At this point, we should be catching, reporting, and notifiying (we're not, see #3864) through the use of celery.

Next stage, if this is going smoothly, is to reduce the reporting logic on the context managers and see if celery can report these more consistently.

Overall, this will reduce the duplication in logic that we're using to report/update builds, and exceptions will be reported in one spot.

0: http://docs.celeryproject.org/en/latest/userguide/tasks.html#on_failure
1: http://docs.celeryproject.org/en/latest/userguide/tasks.html#on_success
2: http://docs.celeryproject.org/en/latest/userguide/tasks.html#on_retry

### LINKED ISSUE #4468: Confusing build failed message (time out and memory consumption)
Recently I have seen some reports of build failure like https://github.com/rtfd/readthedocs.org/issues/4443, where the failed build display two different messages as the failure reason.

Here you can see an example https://readthedocs.org/projects/oapackage/builds/7559891/. The error message is `Build exited due to time out`, but the last command that was executed failed with ` Command killed due to excessive memory consumption`.

### LINKED ISSUE #5672: Unify build_env and setup_env
Currently we are using two different build instances 

https://github.com/rtfd/readthedocs.org/blob/2e797fa1b45fe9772c6cf77c4fd55731d515585d/readthedocs/projects/tasks.py#L454-L454

https://github.com/rtfd/readthedocs.org/blob/2e797fa1b45fe9772c6cf77c4fd55731d515585d/readthedocs/projects/tasks.py#L536

We have some places where we are expecting for the `setup_env`  attr.

This should be only one instance.

### LINKED ISSUE #6452: Initialize all the attributes in build tasks on `__init__` method
Move all the self attributes to the `__init__` method and work from there, right now we initialize the attributes around all the places.

https://github.com/readthedocs/readthedocs.org/blob/c2609f535981e871ff7fe979188cbbe141dd8bc3/readthedocs/projects/tasks.py#L216

https://github.com/readthedocs/readthedocs.org/blob/c2609f535981e871ff7fe979188cbbe141dd8bc3/readthedocs/projects/tasks.py#L301

Ref https://github.com/readthedocs/readthedocs.org/pull/6451#issuecomment-564079573

Edit: I edit the title and the description a little.

### LINKED ISSUE #7587: Notified of a failed build that does not exist
## Details

* Read the Docs project URL: https://readthedocs.org/projects/astropy/
* Build URL (if applicable): https://readthedocs.org/projects/astropy/builds/12149040/
* Read the Docs username (if applicable): pllim

## Expected Result

No notification about a failed build if the build succeeded.

## Actual Result

Got the following email notification but when I checked the build, it was successful (or in a previous occurence, still building). This is not the only time such a situation happened. It has happened with different build IDs in the past few days, but perhaps not for every build?

<hr/>

From: Read the Docs
Sent: Tuesday, October 20, 2020 8:51 PM
Subject: Failed: Astropy (f39a33a3)

Astropy build #12149040

Build Failed for Astropy (latest)

Error: There was a problem with Read the Docs while building your documentation. Please try again later. However, if this problem persists, please report this to us with your build id (12149040).

You can find out more about this failure here:

[Astropy build #12149040](https://readthedocs.org/projects/astropy/builds/12149040/) - failed

If you have questions, a good place to start is the FAQ:

https://docs.readthedocs.io/page/faq.html

You can unsubscribe from these emails in your [Notification Settings](https://readthedocs.org/dashboard/astropy/notifications/)

Keep documenting,
Read the Docs

Read the Docs
https://readthedocs.org

### LINKED ISSUE #7627: Failed build email for successful build
## Details

* Build URL (if applicable): https://readthedocs.org/projects/writethedocs-www/builds/12235484/

## Expected Result

Nothing

## Actual Result

image

### LINKED ISSUE #8762: Log line "Build finished" is not correct
In 

https://github.com/readthedocs/readthedocs.org/blob/637a91ead30496d39c88c50ef34a21c412e4940d/readthedocs/doc_builder/environments.py#L547-L555

we are logging `Build finished` (inside `__exit__`), but that line is called up to 3 times per build because we are using the `DockerBuildEnvironment` (which inherit from `BuildEnvironment`) 3 times with a context manager:

1. when syncing the repository: https://github.com/readthedocs/readthedocs.org/blob/637a91ead30496d39c88c50ef34a21c412e4940d/readthedocs/projects/tasks.py#L354
2. when running the setup: https://github.com/readthedocs/readthedocs.org/blob/637a91ead30496d39c88c50ef34a21c412e4940d/readthedocs/projects/tasks.py#L642
3. when running the build: https://github.com/readthedocs/readthedocs.org/blob/637a91ead30496d39c88c50ef34a21c412e4940d/readthedocs/projects/tasks.py#L642

Note that we are using this log line in the New Relic dashboard and it's currently showing not accurate data.

Related to #3984

### PR DESCRIPTION
**Pull request by humitos**
I'm writing some notes about the decisions I've made, problems I've had and some other concerns and things I'm noticing while working on this refactor. Hopefully, we can come back to these notes in the future to keep cleaning up our code. **These notes are useful for a better understanding of this PR's changes**: https://hackmd.io/4TP3_EizRF-nCDrPvf4_ug

I'm happy to have a call and do some "pair review" for this work because it's quite big. However, I think it won't be too hard to communicate and understand the changes if we are talking to each other.

Also note the _important commits_ are those that are prefixed with `Build:` on their message. So, I suggest taking a look at them mainly.

Closes #3984
Closes #8762
Closes #7627
Closes #7587
Closes #6452
Closes #5672
Closes #4468
It will help to implement #7660

### CODE REVIEW

**Comment by humitos** [comment]
Yay! Only 85 tests skipped 💪🏼 at this point.

**Comment by agjohnson** [comment]
Phew! This seems like a lot of work to wrangling! :muscle:

One thing that would make review easier on the backend team would be splitting up or rebasing some of this work when you're ready. For instance, one commit to split up `project/tasks.py` into multiple files, one commit to alter the files. This would make it much easier to follow along with what has changed in the logic there.

But tests are isolated nicely, and most of the other refactors make sense logically so far. :+1:  It was mostly just the project tasks that were hard to follow here.

File: readthedocs/projects/base.py
Comment (humitos) [inline readthedocs/projects/base.py]: It seems that I committed this file by mistake. It has the same content that `readthedocs/projects/tasks/base.py` --I should have failed with the file rename.

File: readthedocs/projects/tasks/builds.py
Comment (agjohnson) [inline readthedocs/projects/tasks/builds.py:293]: Not sure this is correct, but "NOOP" might not be clear when we are reviewing logs outside the context of code.

```suggestion
            log.warning('Skipping build, build is marked as duplicated.')
```

File: readthedocs/projects/tasks/builds.py
Comment (agjohnson) [inline readthedocs/projects/tasks/builds.py:507]: Correct, we are not, except for historic builds before we introduced BuildCommand -- many years ago at this point. I'd move this discussion to an issue and we can come back to it. I don't know if we've ever come to a conclusion here, but 5 year old builds are almost certainly worthless at this point.

File: readthedocs/projects/tasks/builds.py
Comment (agjohnson) [inline readthedocs/projects/tasks/builds.py:510]: Just curious, I've seen this a few times. Is information on the task that is executing bound somewhere already? That is, will log lines include some metadata on the task and task_id in our logs?

File: readthedocs/projects/tasks/builds.py
Comment (agjohnson) [inline readthedocs/projects/tasks/builds.py:1015]: Just a note here so this doesn't get lost, this todo sounds important.

File: readthedocs/rtd_tests/tests/test_celery.py
Comment (agjohnson) [inline readthedocs/rtd_tests/tests/test_celery.py:49]: Also just a marker to make sure we don't lose this point. This seems important for testing, but I know this is still WIP.

File: readthedocs/rtd_tests/tests/test_doc_building.py
Comment (agjohnson) [inline readthedocs/rtd_tests/tests/test_doc_building.py:38]: Another marker here just to ensure we don't lost this. Another important part that we need before merge.

**Review by agjohnson — COMMENTED** [review]
I jumped around a lot to make sense of everything, and only eyeballed code that looked familiar. I'd take a look through my notes/review, as you might have some more input to add. I left notes so others had some context too.

The logic changes to the build tasks were actually very easy to follow. I love the new code pattern, error handling and just the entire flow of things is way cleaner!

I tried to get some diffs out of the changes to isolate the logic that changed in moved/refactored code, but was having trouble following where code was moved to. I was eyeballing more green code than I'd like for such a large change, so I can't honestly say I was able to give good review to code that was moved -- build environment and command tests, and the projects tasks code besides some of the main tasks.

A couple things stood out:

- This PR is saving state to Task instances. This has historically caused fairly severe bugs, especially if there is parallel task execution on the worker. This would be a pattern to discuss in our meeting.
- We should definitely review all of the TODO notes in the diff here before merging. Just to make sure we didn't miss anything important. It might be a good idea to surface some as later issues. This PR still seems a bit of WIP, so I stopped noting TODOs at some point.

File: readthedocs/core/urls/__init__.py
Comment (humitos) [reply readthedocs/core/urls/__init__.py:20]: This is part of a decision that I've made to delete related non-used/invalid code together with these changes. The wipe "feature" is not a feature anymore because we do a full clean after the build is finished (`RTD_CLEAN_AFTER_BUILD=True`). So, wiping the environment has no effect and it was a confusing action for our users.

File: readthedocs/core/utils/__init__.py
Comment (humitos) [reply readthedocs/core/utils/__init__.py:80]: I removed these two.

- `record` was removed from a the top level (when triggering a task) because there is no use case for that. We _always_ want to record a `Build` object in our database if there is a build happening. However, I kept it at `BuildCommand` level, because there are commands that we want to execute in the environment without showing them to the user. In that case, we are using `record=False`
- `force` stopped making sense when we started using `RTD_CLEAN_AFTER_BUILD=True`. With this setting, we are always "forcing a build" since there is no cached environment to re-use.

I'm happy to bring these features back (including the "cache whole build environment as `.tar.gz` on S3") when we have decided to start using them. I didn't want to refactor/maintain all that code that it's not currently used because it added a lot of extra work. I think that re-add them when required will be easier and cleaner with the new structure.

File: readthedocs/projects/tests/mockers.py
Comment (humitos) [reply readthedocs/projects/tests/mockers.py:12]: This class is the most complex thing from this PR, unfortunately. It's _really_ hard to test our code because it requires a remote server running an API _and_ Docker to execute the commands.

This code is the most accurate attempt that I was able to do to mock these calls in a granular way making sure that we can execute as much code as possible. However, it's not perfect and there are inner functions that I wasn't able to mock properly yet. For example, I wrote this problem for this case in https://github.com/readthedocs/readthedocs.org/blob/d5a8ffb0e458e447772bc7d6008e5a39893f3af7/readthedocs/projects/tests/mockers.py#L72-L79

This class is similar to the work tried in `EnvironmentMockGroup` at https://github.com/readthedocs/readthedocs.org/blob/d5a8ffb0e458e447772bc7d6008e5a39893f3af7/readthedocs/rtd_tests/mocks/environment.py#L7 but covers all the build's process. The new class should replace the old one.

I'm open to receiving feedback here and finding out a better way to do this. I had a really hard time dealing with mocks and trying to make the full build process "runable" from a test.

File: readthedocs/projects/tests/test_build_tasks.py
Comment (humitos) [reply readthedocs/projects/tests/test_build_tasks.py:1]: This is a new file mostly written from scratch. Some of these tests are completely new and some of them are migrated to use the new mocking system from other files. However, they were completely rewritten. In most cases, only the name kept the same 😄 

Basically, I migrated the "the logic's test case" (e.g. "check that this config file executes this command") but not its code.

File: readthedocs/doc_builder/environments.py
Comment (humitos) [reply readthedocs/doc_builder/environments.py:550]: They can be completely removed. I wasn't sure if they were needed and that's why I commented them out initially, but we are not using this code anymore.

File: readthedocs/doc_builder/environments.py
Comment (humitos) [reply readthedocs/doc_builder/environments.py:641]: Yes, this code was moved to the task itself.

The general rule behind this was: "The task should be in charge of everything related to updating the build/version/project. On the other hand, the environment should only be in charge of running the commands + hitting the API with the command's output because it's the only one that has access to that data"

Comment (humitos) [reply readthedocs/doc_builder/environments.py:641]: By moving this logic to the task, it's _always_ clear _when_ we need to update the build and _how_ depending on where we are in the flow (e.g. `after_return` the build is FINISHED; while installing dependencies, the build is INSTALLING, etc). The logic got simplified a lot.

Besides, we always update the build in the same way: update the `self.build` dictionary and call `self.update_build` from the task. This also may reduce the number of hits to the API required (e.g. we could update the `self.build` dictionary _without_ hitting the API immediately if required).

This is also useful together with `on_success`, `on_failure` and `after_return` where the first two methods update the `self.build['success']` with a specific value, and then in `after_return` we hit the API _without knowing_ if the build was successful or not.

File: readthedocs/doc_builder/exceptions.py
Comment (humitos) [reply readthedocs/doc_builder/exceptions.py:38]: He, yeah, I'm not proud of my copy here 😓 . I wrote this just to demonstrate that now we can differentiate "application errors" (mainly _our_ fault: couldn't connect to docker, API from webs failed, etc) from "user errors" (mainly their fault: invalid configuration, invalid submodule, etc)

- "Application errors" will show the generic message: "Something happens and we don't know what it was. If the problem persists, contact us".
- "User errors" will show a message telling the user there is something wrong on their side: "A command failed and this is probably something that you can solve on your own. It was not an internal Read the Docs issue"

File: readthedocs/projects/tasks/__init__.py
Comment (humitos) [reply readthedocs/projects/tasks/__init__.py:9]: Actually, I realized this when I spun up an instance with this code. `webs` are currently triggering `readthedocs.projects.tasks.update_docs_task` which was not registered in my branch. Adding this code, solved that problem.

However, when we finally deploy this, if we deploy `webs` first they will start triggering task named as `readthedocs.projects.tasks.builds.update_docs_task` and old instances of builders won't have it registered. So, we should probably deploy builders first 🤔  --or overwrite the names of the task to keep the old name _and_ the new name at the same time.

File: readthedocs/projects/tasks/__init__.py
Comment (humitos) [reply readthedocs/projects/tasks/__init__.py:24]: I'm not sure to understand this comment. There is nothing we can do now about this.

Note that we are not upgrading Celery on this PR. Because of that, we need to emulate a behavior from a future release: calling `before_start` method by hand.

File: readthedocs/projects/tasks/builds.py
Comment (humitos) [reply readthedocs/projects/tasks/builds.py:156]: "Yes" to everything you noted here.

I ended up with this ugly hack for now because I didn't find a way to be able to use all the Celery machinery of the handlers by using `TaskStep` idea. I'm happy to get some help on this if we can find a better way to solve this problem.

It may not be hard, but what I've tried didn't work at first and I wanted to continue moving forward instead of being block immediately when I started the refactor 😊 

Also, note that we are not executing tasks in parallel now. Our builders only run 1 concurrent build task.

File: readthedocs/projects/tasks/builds.py
Comment (humitos) [reply readthedocs/projects/tasks/builds.py:241]: I'm not taking them in this PR, no. These are notes that I found useful from the documentation while reading it and that could apply to our use cases. I'm not sure they are _good_, tho. We can come back to them later and check if it's worth the effort.

File: readthedocs/projects/tasks/builds.py
Comment (humitos) [reply readthedocs/projects/tasks/builds.py:507]: OK, I think we can delete these updates then and just use the default from the database. In the future, we can even remove the db fields.

File: readthedocs/projects/tasks/builds.py
Comment (humitos) [reply readthedocs/projects/tasks/builds.py:510]: Yes! This is the magic of `log.bind` from `structlog` (see https://www.structlog.org/en/stable/getting-started.html#building-ctx)

We are calling `log.bind` on `before_start` with all the useful data for this particular build: https://github.com/readthedocs/readthedocs.org/blob/d5a8ffb0e458e447772bc7d6008e5a39893f3af7/readthedocs/projects/tasks/builds.py#L329-L336

The following `log.info` will include all those keys + the message we are logging.

Besides, there are some "extra Celery metadata" that is added by the Django app `django-structlog`. For example, `task_id` (see https://github.com/jrobichaud/django-structlog/blob/master/django_structlog/celery/receivers.py)

File: readthedocs/projects/tasks/builds.py
Comment (humitos) [reply readthedocs/projects/tasks/builds.py:1015]: I did this temporarily. Basically, I want to have required arguments _being required_ instead of optional keyword arguments. They are mainly `version_pk` and `build_pk`. So, to avoid changing the signature of the task and breaking a lot of code immediately when I started the refactor, I did this hack for now.

File: readthedocs/projects/tasks/__init__.py
Comment (humitos) [reply readthedocs/projects/tasks/__init__.py:9]: Actually, I think there may be other tasks that will be affected by this. Those that are executed by `web-celery` as well. I'm thinking on `fileify` too.

Comment (humitos) [reply readthedocs/projects/tasks/__init__.py:9]: https://docs.celeryproject.org/en/stable/userguide/tasks.html#names

File: readthedocs/doc_builder/environments.py
Comment (humitos) [inline readthedocs/doc_builder/environments.py:682]: We are not canceling the build on these cases.

File: readthedocs/projects/tasks/builds.py
Comment (ericholscher) [inline readthedocs/projects/tasks/builds.py:116]: This could be:

`self.data[task_id].version`.

File: readthedocs/projects/tasks/builds.py
Comment (humitos) [inline readthedocs/projects/tasks/builds.py:306]: This hack should probably solved by namespacing our own data and just delete this only attribute when cleaning (`after_return`)

File: readthedocs/projects/tasks/builds.py
Comment (agjohnson) [reply readthedocs/projects/tasks/builds.py:156]: Notes from meeting on a path forward perhaps worth exploring here:

* Move our attributes from `self` to a namespaced `self.data = {}`. Don't manipulate attributes at `self` on `Task`
* Or, consider usage of the TaskStep pattern to encapsulate data. This probably still needs a namespace and dictionary lookup
* Maybe use `task_id` or build pk from args for `self.data` key lookup
* We might need this to be a cache-like object with TTL or it could produce a memory leak as builds accumulate. Deleting keys from the data namespace should be fairly accurate however.

```
task_data = self.data[task_id]
# All of the attributes on `self` on `Task` are in `self.data[x]` instead
build = task_data.build
```

File: readthedocs/projects/tasks/builds.py
Comment (agjohnson) [reply readthedocs/projects/tasks/builds.py:507]: I've so far skipped display of these types of builds with the new templates. The old display was just several blocks of stdout/stderr, not sequential command and command output, so this is a big pain to support in a nicer design. I'm +1 on dealing with this in a permanent way.

**Comment by agjohnson** [comment]
Just a note for review purposes later. At the time of our meeting, the latest commit was: d5a8ffb

File: readthedocs/projects/tasks/__init__.py
Comment (humitos) [reply readthedocs/projects/tasks/__init__.py]: I updated the code in this commit a393ddbb664c25b9e211517d5049e5f2297f4959. Let me know if I was clear enough and if it does make sense to you.

**Comment by humitos** [comment]
@readthedocs/backend this work is ready for review. I think I cleaned up all the things that were missing after our meeting. There are still some tests skipped (~8) but I think it's fine to not block this PR on that and try to keep moving forward. Adding more tests to the full build process will be ongoing work that we will need to do. There were, and still are, plenty of code that it's not _well-tested_, but I want to come back in the near future to continue with them.

I also upgraded the code running in my "testing in production" instance that we can use to test some more projects. Just assigning `Project.build_queue = 'celery:handlers'` should be enough. Let me know if you find any issues or you have feedback to share.

Remember that we are planning to deploy this work on Feb 8th.

File: readthedocs/core/urls/__init__.py
Comment (agjohnson) [reply readthedocs/core/urls/__init__.py:20]: Makes sense! I'm +1 on the removal in this PR. I've also removed this from the new site theme as well.

File: readthedocs/core/utils/__init__.py
Comment (agjohnson) [reply readthedocs/core/utils/__init__.py:80]: This makes sense, I agree with your interpretation. I was originally looking for the BuildCommand record, but was just confused by the same argument here.

File: readthedocs/projects/tests/mockers.py
Comment (agjohnson) [reply readthedocs/projects/tests/mockers.py:12]: We went over this in the meeting and all my questions so far were covered :+1: 

I liked the approach of testing the model updates through the API. This is more helpful that testing model state at the begin/end of a test. I also appreciate how it is closer to an integration test in that we're not testing each individual command executions, but *the outcome* from executing commands and flow through our build process.

As with any mocking, it's hard to track what is not mocked, but we (probably more like developers in general) have yet to come up with a good answer to that mystery.

File: readthedocs/doc_builder/environments.py
Comment (agjohnson) [reply readthedocs/doc_builder/environments.py:641]: Some of the tasks and environment changes I still want to spend some time with and digest, but so far the changes to the pattern here make sense to me, and I get how environment interacts with the build task at a high level.

File: readthedocs/projects/tasks/__init__.py
Comment (agjohnson) [reply readthedocs/projects/tasks/__init__.py]: Aaahhh, okay that makes more sense. I read this as "We need to upgrade to 5.2.x for this method to be used by celery"

File: readthedocs/projects/tasks/builds.py
Comment (agjohnson) [reply readthedocs/projects/tasks/builds.py]: It might be worth noting this somewhere, perhaps just start an issue preemptively? We'll have some issues come out of the left over TODOs here at the end I'm sure, but don't need to track every little change.

File: readthedocs/projects/tasks/__init__.py
Comment (agjohnson) [reply readthedocs/projects/tasks/__init__.py]: The code mostly makes sense, but I might be a little lost now on what we would need to do to test this. It seems this commit solves the issue of task renaming entirely for now, correct? That is, to start an instances with this branch on a new instance, there is no special set up required?

File: readthedocs/projects/tasks/builds.py
Comment (agjohnson) [reply readthedocs/projects/tasks/builds.py:467]: :magic_wand:

File: readthedocs/projects/tasks/__init__.py
Comment (humitos) [reply readthedocs/projects/tasks/__init__.py]: No, I solved it. The instance is running code now and attending tasks triggered by "old webs instances" using the "old task name". When they receive it, they re-trigger the task using the new name and passing the proper parameters (in case the signature has changed)

However, note that the commit that I linked was not correct and the newer solution was a little different. See these commits: 9154e8d363bc6843697a38268f8d51c81bb44bb4 and 1e9c7240251ba0e67c2945da07172804b2ee32cc

File: readthedocs/projects/tasks/builds.py
Comment (humitos) [reply readthedocs/projects/tasks/builds.py]: Yeah. I plan to come back to the notes linked in the description and open the corresponding issues.

**Review by agjohnson — COMMENTED** [review]
I reviewed the changes that happened since our meeting, looking good still! I won't yet call this a +1 yet, but this is mostly me understanding the changes still.

Things I haven't reviewed well yet, and where I'd like to spend some time in the next few days:

- Some of the code changes that happened before our meeting, and reviewing the project/build tasks in a bit more depth. I mostly eyeballed the difficult files to understand the logic, but didn't look very deeply at the changes yet. I'm guessing it might help to jump on another call at some point to discuss any remaining questions here, but let me see what questions I come up with before I distract you with this.
- Skipped tests and test coverage. I'm still wrapping my head around some of the changes still, but I don't see anything really needing changes. Some of the skipped tests might worry me because I don't trust I can help discover issues more than good test coverage would. I do agree with your instinct on some of the skipped tests so far though, so most likely nothing will come of this.

I'm planning on spending more time with this in the next few days.

File: readthedocs/projects/tasks/builds.py
Comment (humitos) [reply readthedocs/projects/tasks/builds.py:140]: I created a "namespace" to store all the task-related data under `self.data` and do a cleanup each time a new task is executed. This way, we will avoid the problem of shared data between different execution tasks.

Please, take a look at it at https://github.com/readthedocs/readthedocs.org/pull/8874 (I created a different PR to make it easier to review, but basically is a search&replace of `self.` by `self.data.`)

Let me know if the new pattern is aligned with the ideas that we discussed in the meeting.

**Comment by humitos** [comment]
I'm not able to solve the `cyclic-import`. The module reported with the cyclic import by pylint does not make any sense. So, I'm not sure how to solve this. I spent some time trying to figure it out, but I failed. It seems to be a bug in pylint or something like that 🤷🏼 

I opened https://github.com/readthedocs/common/pull/103 to disable the `cycle-import` check in case we want to follow that path because I ran out of ideas about what it could be and the code is working properly locally and in production (with the instance deployed with this branch).

**Comment by agjohnson** [comment]
pycycle is equally not helpful here :disappointed: 

image

(the cyclic import details are missing, this should tell us where the cyclic import is)

**Comment by agjohnson** [comment]
Tuning pycycle a bit more to narrow this down, the error goes away:

```
(rtfd38) community/ humitos/celery-handlers* 💀  pycycle --here --ignore user_builds,bower_components,docs,media,common,.ropeproject,dockerfiles,node_modules          
Project successfully transformed to AST, checking imports for cycles..
No worries, no cycles here!
If you think some cycle was missed, please open an Issue on Github.
Finished.
```

:shrug::question:

File: readthedocs/projects/tasks/builds.py
Comment (agjohnson) [reply readthedocs/projects/tasks/builds.py:140]: Yeah, this was the first step that we talked about, and I'm happy to have it included in this initial batch of work.

We do need to think of ways to make this thread safe however, but we can come back to that after merging this. I don't think we were too concerned about complicating the implementation here for this use case.

**Review by agjohnson — APPROVED** [review]
I finally got a chance to spend some more time with this PR. I reviewed work since our meeting and re-reviewed project tasks modules that I previously skimmed.

At a high level, all of the changes make sense and the new celery task pattern is really easy to follow. I can't say that I did a great job reviewing the code at a low level though. There was a lot of code that was moved alongside refactoring.

So, this is a +1, but also a little bit of a yolo +1 as well. I think we are mostly confident at this point that no obvious issues were introduced here, but we can probably expect a PR of this size to introduce a few non-obvious issues as well.

@humitos does have notes accumulated here on next steps. We should turn some into issues to return to later.

**Pull request merged**
